File: news-social-intelligence-data-pipeline/social_intel/x_twitter.py
"""X/Twitter scraper using xurl CLI (requires OAuth2 setup).

Uses the official xurl CLI for X API v2 access.
Requires: xurl auth setup (user must do this manually)

Output format matches our normalized article schema.

IMPORTANT: User must complete OAuth2 setup manually:
1. Create app at https://developer.x.com/en/portal/dashboard
2. xurl auth apps add my-app --client-id YOUR_ID --client-secret YOUR_SECRET
3. xurl auth oauth2 --app my-app
4. xurl auth default my-app
"""

import json
import subprocess
import sys
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from .lib import combined_relevance

XURL_TIMEOUT = 30

# Depth-aware limits
DEPTH_LIMITS = {
    "quick": 10,
    "default": 25,
    "deep": 50,
}


def _log(msg: str) -> None:
    sys.stderr.write(f"[X/Twitter] {msg}\n")
    sys.stderr.flush()


def _run_xurl(args: List[str], timeout: int = XURL_TIMEOUT) -> Optional[Dict]:
    """Run xurl command and return parsed JSON. Never raises."""
    cmd = ["xurl"] + args
    
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        if result.returncode == 0 and result.stdout:
            return json.loads(result.stdout)
        else:
            if "No apps registered" in result.stdout or "No apps registered" in result.stderr:
                _log("Auth not configured - user must run xurl auth setup")
            else:
                _log(f"xurl error: {result.stderr[:200] if result.stderr else result.stdout[:200]}")
            return None
    except subprocess.TimeoutExpired:
        _log(f"xurl timeout after {timeout}s")
        return None
    except FileNotFoundError:
        _log("xurl not found - install with: curl -fsSL https://raw.githubusercontent.com/xdevplatform/xurl/main/install.sh | bash")
        return None
    except json.JSONDecodeError as e:
        _log(f"JSON parse error: {e}")
        return None
    except Exception as e:
        _log(f"xurl error: {e}")
        return None


def _check_auth() -> bool:
    """Check if xurl auth is configured. Never raises."""
    result = _run_xurl(["auth", "status"])
    if result is None:
        return False
    # If we got valid JSON back, auth is likely configured
    return True


def _parse_tweet(tweet: Dict[str, Any], query: str = "") -> Optional[Dict[str, Any]]:
    """Parse X API tweet into normalized article dict. Never raises."""
    try:
        tweet_id = tweet.get("id", "")
        text = tweet.get("text", "")
        
        if not tweet_id or not text:
            return None
        
        # Calculate relevance
        relevance = combined_relevance(query, text, "") if query else 0.0
        
        # Parse created_at (ISO format from X API v2)
        created_at = tweet.get("created_at")
        date_str = None
        created_utc = None
        if created_at:
            try:
                dt = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
                date_str = dt.date().isoformat()
                created_utc = dt.timestamp()
            except Exception:
                pass
        
        # Extract author info
        author = tweet.get("author_id", "")
        
        # Build URL
        url = f"https://x.com/i/status/{tweet_id}"
        
        # Get metrics
        public_metrics = tweet.get("public_metrics", {})
        
        return {
            "id": tweet_id,
            "title": text[:100] + ("..." if len(text) > 100 else ""),
            "url": url,
            "description": text,
            "source": "X/Twitter",
            "author": author,
            "score": public_metrics.get("like_count", 0),
            "num_comments": public_metrics.get("reply_count", 0),
            "created_utc": created_utc,
            "date": date_str,
            "relevance": round(relevance, 3),
            "why_relevant": "xurl search",
            "collected_at": datetime.now(timezone.utc).isoformat(),
            "content_type": "x_twitter",
            "metadata": {
                "tweet_id": tweet_id,
                "retweet_count": public_metrics.get("retweet_count", 0),
                "like_count": public_metrics.get("like_count", 0),
                "reply_count": public_metrics.get("reply_count", 0),
                "quote_count": public_metrics.get("quote_count", 0),
                "impression_count": public_metrics.get("impression_count", 0),
            },
        }
    except Exception as e:
        _log(f"Parse error: {e}")
        return None


def is_available() -> bool:
    """Check if X/Twitter scraping is available (xurl installed + auth configured)."""
    try:
        result = subprocess.run(
            ["xurl", "auth", "status"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        return "No apps registered" not in result.stdout and result.returncode == 0
    except Exception:
        return False


def search_tweets(
    query: str,
    depth: str = "default",
) -> List[Dict[str, Any]]:
    """Search X/Twitter posts. Never raises.
    
    Args:
        query: Search query string (supports X search operators)
        depth: 'quick', 'default', or 'deep' - controls result limit
        
    Returns:
        List of normalized article dicts, empty on failure or if auth not configured
    """
    if not is_available():
        _log("X/Twitter not available - xurl auth not configured")
        return []
    
    limit = DEPTH_LIMITS.get(depth, DEPTH_LIMITS["default"])
    
    _log(f"Searching: {query} (depth={depth}, limit={limit})")
    
    data = _run_xurl(["search", query, "-n", str(limit)])
    
    if not data:
        return []
    
    # X API v2 returns data in {"data": [...]} format
    tweets = data.get("data", [])
    if isinstance(data, list):
        tweets = data
    
    results = []
    for tweet in tweets:
        parsed = _parse_tweet(tweet, query)
        if parsed:
            results.append(parsed)
    
    _log(f"Found {len(results)} results")
    return results


def get_user_timeline(
    username: str,
    limit: int = 20,
) -> List[Dict[str, Any]]:
    """Get recent tweets from a user. Never raises.
    
    Args:
        username: X/Twitter username (with or without @)
        limit: Max tweets to return
        
    Returns:
        List of normalized article dicts, empty on failure
    """
    if not is_available():
        _log("X/Twitter not available - xurl auth not configured")
        return []
    
    username = username.lstrip("@")
    _log(f"Getting timeline for @{username}")
    
    # First get user info
    user_data = _run_xurl(["user", username])
    if not user_data:
        return []
    
    # Get user's tweets via raw API
    user_id = user_data.get("data", {}).get("id")
    if not user_id:
        return []
    
    data = _run_xurl([f"/2/users/{user_id}/tweets", "-n", str(limit)])
    
    if not data:
        return []
    
    tweets = data.get("data", [])
    results = []
    
    for tweet in tweets:
        parsed = _parse_tweet(tweet)
        if parsed:
            parsed["author"] = username
            results.append(parsed)
    
    _log(f"Found {len(results)} tweets")
    return results


def get_mentions(limit: int = 20) -> List[Dict[str, Any]]:
    """Get mentions of the authenticated user. Never raises.
    
    Args:
        limit: Max mentions to return
        
    Returns:
        List of normalized article dicts, empty on failure
    """
    if not is_available():
        _log("X/Twitter not available - xurl auth not configured")
        return []
    
    _log(f"Getting mentions (limit={limit})")
    
    data = _run_xurl(["mentions", "-n", str(limit)])
    
    if not data:
        return []
    
    tweets = data.get("data", [])
    results = []
    
    for tweet in tweets:
        parsed = _parse_tweet(tweet)
        if parsed:
            results.append(parsed)
    
    _log(f"Found {len(results)} mentions")
    return results


# Search operators cheatsheet
SEARCH_OPERATORS = """
X/Twitter Search Operators:
- from:username - tweets from user
- to:username - replies to user
- @username - mentions of user
- #hashtag - hashtag search
- "exact phrase" - exact match
- lang:en - language filter
- -filter:retweets - exclude retweets
- filter:media - only with media
- min_retweets:N - minimum retweets
- min_faves:N - minimum likes
- since:YYYY-MM-DD - after date
- until:YYYY-MM-DD - before date
"""


if __name__ == "__main__":
    print("=== X/Twitter Status ===")
    if is_available():
        print("✅ xurl auth configured")
        
        print("\n=== Search: AI ===")
        results = search_tweets("AI", depth="quick")
        for item in results[:3]:
            print(f"  [{item['score']}] {item['title'][:60]}")
    else:
        print("❌ xurl auth not configured")
        print("\nTo set up:")
        print("1. Create app at https://developer.x.com/en/portal/dashboard")
        print("2. xurl auth apps add my-app --client-id YOUR_ID --client-secret YOUR_SECRET")
        print("3. xurl auth oauth2 --app my-app")
        print("4. xurl auth default my-app")

File: news-social-intelligence-data-pipeline/src/social/x_twitter.rs
//! X/Twitter collector via xurl CLI (requires OAuth2 setup).
//!
//! Uses the official xurl CLI for X API v2 access.
//! User must complete OAuth2 setup manually before this works:
//! 1. Create app at https://developer.x.com/en/portal/dashboard
//! 2. `xurl auth apps add my-app --client-id YOUR_ID --client-secret YOUR_SECRET`
//! 3. `xurl auth oauth2 --app my-app`
//! 4. `xurl auth default my-app`

use std::process::Command;
use tracing::{info, warn};
use chrono::Utc;

use super::SocialArticle;
use super::relevance::combined_relevance;

/// Timeout for xurl subprocess calls (reserved for future use with process timeout).
#[allow(dead_code)]
const XURL_TIMEOUT_SECS: u64 = 30;

/// Run xurl with the given arguments and return stdout on success.
/// Returns `None` on any error (binary not found, timeout, non-zero exit).
fn run_xurl(args: &[&str]) -> Option<String> {
    let output = match Command::new("xurl")
        .args(args)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()
    {
        Ok(o) => o,
        Err(e) => {
            warn!("xurl not found or failed to spawn: {}", e);
            return None;
        }
    };

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        let stdout = String::from_utf8_lossy(&output.stdout);
        if stdout.contains("No apps registered") || stderr.contains("No apps registered") {
            warn!("xurl auth not configured - user must run xurl auth setup");
        } else {
            warn!("xurl exited with status {}: {}", output.status, &stderr[..stderr.len().min(200)]);
        }
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    if stdout.is_empty() {
        None
    } else {
        Some(stdout)
    }
}

/// Parse a tweet JSON value into a normalized SocialArticle.
/// Returns `None` if required fields (id, text) are missing.
fn parse_tweet(tweet: &serde_json::Value, query: &str) -> Option<SocialArticle> {
    let tweet_id = tweet.get("id")?.as_str()?;
    let text = tweet.get("text")?.as_str()?;

    if tweet_id.is_empty() || text.is_empty() {
        return None;
    }

    // Title = first 100 chars + "..." if truncated
    let title = if text.len() > 100 {
        let boundary = text
            .char_indices()
            .nth(100)
            .map(|(i, _)| i)
            .unwrap_or(text.len());
        format!("{}...", &text[..boundary])
    } else {
        text.to_string()
    };

    let relevance = if query.is_empty() {
        0.0
    } else {
        combined_relevance(query, text, "", 0.7)
    };

    // Parse created_at (ISO 8601 format from X API v2)
    let created_at_str = tweet.get("created_at").and_then(|v| v.as_str());
    let (created_utc, date) = if let Some(ts) = created_at_str {
        let parsed = chrono::DateTime::parse_from_rfc3339(ts)
            .or_else(|_| chrono::DateTime::parse_from_rfc3339(&ts.replace('Z', "+00:00")))
            .ok();
        match parsed {
            Some(dt) => (Some(dt.timestamp() as f64), Some(dt.format("%Y-%m-%d").to_string())),
            None => (None, None),
        }
    } else {
        (None, None)
    };

    let author_id = tweet
        .get("author_id")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let url = format!("https://x.com/i/status/{}", tweet_id);

    // Extract public metrics
    let public_metrics = tweet.get("public_metrics");
    let like_count = public_metrics
        .and_then(|m| m.get("like_count"))
        .and_then(|v| v.as_i64())
        .unwrap_or(0);
    let reply_count = public_metrics
        .and_then(|m| m.get("reply_count"))
        .and_then(|v| v.as_i64())
        .unwrap_or(0);
    let retweet_count = public_metrics
        .and_then(|m| m.get("retweet_count"))
        .and_then(|v| v.as_i64())
        .unwrap_or(0);

    let metadata = serde_json::json!({
        "tweet_id": tweet_id,
        "retweet_count": retweet_count,
        "like_count": like_count,
        "reply_count": reply_count,
    });

    Some(SocialArticle {
        id: tweet_id.to_string(),
        title,
        url,
        description: text.to_string(),
        source: "X/Twitter".to_string(),
        author: author_id,
        score: like_count,
        num_comments: reply_count,
        created_utc,
        date,
        relevance,
        collected_at: Utc::now().to_rfc3339(),
        content_type: "x_twitter".to_string(),
        metadata,
    })
}

/// Check if xurl CLI is installed and auth is configured.
/// Returns `true` if `xurl auth status` exits 0 and output does not contain
/// "No apps registered".
pub fn is_available() -> bool {
    let output = match Command::new("xurl")
        .args(["auth", "status"])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()
    {
        Ok(o) => o,
        Err(_) => return false,
    };

    if !output.status.success() {
        return false;
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    !stdout.contains("No apps registered")
}

/// Search X/Twitter for tweets matching the query.
/// Returns an empty Vec if xurl is not available or on any error.
pub fn search_tweets(query: &str, limit: usize) -> Vec<SocialArticle> {
    if !is_available() {
        warn!("X/Twitter not available - xurl auth not configured");
        return Vec::new();
    }

    info!("🐦 X/Twitter search: {} (limit={})", query, limit);

    let limit_str = limit.to_string();
    let output = match run_xurl(&["search", query, "-n", &limit_str]) {
        Some(o) => o,
        None => return Vec::new(),
    };

    let parsed: serde_json::Value = match serde_json::from_str(&output) {
        Ok(v) => v,
        Err(e) => {
            warn!("Failed to parse xurl JSON: {}", e);
            return Vec::new();
        }
    };

    // X API v2 returns {"data": [...]} or a plain array
    let tweets = if let Some(arr) = parsed.get("data").and_then(|d| d.as_array()) {
        arr.clone()
    } else if let Some(arr) = parsed.as_array() {
        arr.clone()
    } else {
        return Vec::new();
    };

    let mut results = Vec::new();
    for tweet in &tweets {
        if let Some(article) = parse_tweet(tweet, query) {
            results.push(article);
        }
    }

    info!("🐦 X/Twitter found {} results", results.len());
    results
}

/// Get recent tweets from a user's timeline.
/// Returns an empty Vec if xurl is not available or on any error.
pub fn get_user_timeline(username: &str, limit: usize) -> Vec<SocialArticle> {
    if !is_available() {
        warn!("X/Twitter not available - xurl auth not configured");
        return Vec::new();
    }

    let username = username.trim_start_matches('@');
    info!("🐦 X/Twitter timeline: @{} (limit={})", username, limit);

    // First get user info to obtain user ID
    let user_output = match run_xurl(&["user", username]) {
        Some(o) => o,
        None => return Vec::new(),
    };

    let user_data: serde_json::Value = match serde_json::from_str(&user_output) {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };

    let user_id = user_data
        .get("data")
        .and_then(|d| d.get("id"))
        .and_then(|id| id.as_str());

    let user_id = match user_id {
        Some(id) => id.to_string(),
        None => return Vec::new(),
    };

    // Get user's tweets
    let endpoint = format!("/2/users/{}/tweets", user_id);
    let limit_str = limit.to_string();
    let output = match run_xurl(&[&endpoint, "-n", &limit_str]) {
        Some(o) => o,
        None => return Vec::new(),
    };

    let parsed: serde_json::Value = match serde_json::from_str(&output) {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };

    let tweets = if let Some(arr) = parsed.get("data").and_then(|d| d.as_array()) {
        arr.clone()
    } else if let Some(arr) = parsed.as_array() {
        arr.clone()
    } else {
        return Vec::new();
    };

    let mut results = Vec::new();
    for tweet in &tweets {
        if let Some(mut article) = parse_tweet(tweet, "") {
            article.author = username.to_string();
            results.push(article);
        }
    }

    info!("🐦 X/Twitter timeline found {} tweets", results.len());
    results
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_tweet_valid() {
        let tweet = serde_json::json!({
            "id": "1234567890",
            "text": "Artificial intelligence is transforming the world of finance and technology.",
            "created_at": "2025-06-10T12:00:00+00:00",
            "author_id": "987654321",
            "public_metrics": {
                "like_count": 42,
                "reply_count": 5,
                "retweet_count": 10
            }
        });

        let article = parse_tweet(&tweet, "artificial intelligence").unwrap();

        assert_eq!(article.id, "1234567890");
        assert_eq!(article.url, "https://x.com/i/status/1234567890");
        assert_eq!(article.source, "X/Twitter");
        assert_eq!(article.content_type, "x_twitter");
        assert_eq!(article.author, "987654321");
        assert_eq!(article.score, 42);
        assert_eq!(article.num_comments, 5);
        assert_eq!(article.date, Some("2025-06-10".to_string()));
        assert!(article.relevance > 0.0);

        // Check metadata
        let meta = &article.metadata;
        assert_eq!(meta["tweet_id"], "1234567890");
        assert_eq!(meta["retweet_count"], 10);
        assert_eq!(meta["like_count"], 42);
        assert_eq!(meta["reply_count"], 5);
    }

    #[test]
    fn test_is_available_returns_false_when_not_installed() {
        // On most CI/dev machines without xurl, this returns false.
        // On machines with xurl configured, it may return true.
        // We verify it doesn't panic regardless.
        let result = is_available();
        let _ = result;
    }

    #[test]
    fn test_parse_tweet_missing_text() {
        let tweet = serde_json::json!({
            "id": "111222333",
            "author_id": "444555666"
        });

        let result = parse_tweet(&tweet, "test");
        assert!(result.is_none());
    }
}
